### DISCUSSION TITLE
How to place a buffer in OCRAM region of memory on NXP MIMXRT1062 processor

### DISCUSSION THREAD

**Discussion by asteriskSF** [zephyrproject-rtos/zephyr#69886] [Q&A] [answered]
Zephyr 3.5.0
MIMXRT1062 processor

I have an project where the application runs completely from RAM.  The processor provides 3 distinct regions ITCM, DTCM and OCRAM.  The OCRAM is best for DMA operation and by default occupies 768k out of 1024k.   I will be receiving large amounts of DATA from LPSPI using DMA so placing these buffers would be perfect for OCRAM.   I also would like to allocate 2 buffers which I alternate between to avoid race conditions between receiving and the processing, though I doubt this matters to the implementation.

How can I annotate a memory buffer so that it will automatically be placed in the OCRAM region?

@DerekSnell perhaps you can suggest a solution?

**Comment by DerekSnell** [comment, marked as answer]
Hi @asteriskSF ,
One way to do this is to follow the [api_ext test](https://github.com/zephyrproject-rtos/zephyr/tree/main/tests/lib/devicetree/api_ext).  

In the devicetree, add the property `zephyr,memory-region` to the memory node you want data placed in, like in [app.overlay](https://github.com/zephyrproject-rtos/zephyr/blob/main/tests/lib/devicetree/api_ext/app.overlay).  And then in your declaration, add the attribute `Z_GENERIC_SECTION(LINKER_DT_NODE_REGION_NAME(DT_NODELABEL(<node_label_name>)))`, one example of this is in the [memc test](https://github.com/zephyrproject-rtos/zephyr/blob/9fa4cd8226ced9ecfcfb34bf0aa5e86d1abceb1f/tests/drivers/memc/ram/src/main.c#L15C2-L15C68).

Also, if using those macros, be sure to `#include <zephyr/linker/devicetree_regions.h>`.  I learned that leasson the hard way :)

Hope that helps

**Comment by asteriskSF** [comment]
It appears the OCRAM already has the memory region labels added:

zephyr/dts/arm/nxp/nxp_rt10xx.dtsi  has:
			ocram: ocram@20200000 {
				compatible = "zephyr,memory-region", "mmio-sram";
				reg = <0x20200000 DT_SIZE_K(256)>;
				zephyr,memory-region = "OCRAM";
			}

and zephyr/dts/arm/nxp/nxp_rt1060.dtsi  has:
&ocram {
	reg = <0x20200000 DT_SIZE_K(768)>;
};

so I made driver declaration like this:
#define MY_DRIVER_INIT(inst)                                                    \
    static frame_buffer_t frame_buffers_##inst[NUM_FRAME_BUFFERS]                \
	          Z_GENERIC_SECTION(LINKER_DT_NODE_REGION_NAME(DT_NODELABEL(ocram)));

and it was placed in OCRAM!
[7/7 100% :: 3.098] Linking C executable zephyr\zephyr.elf
Memory region         Used Size  Region Size  %age Used
           SDRAM:          0 GB        32 MB      0.00%
           FLASH:       65964 B       128 KB     50.33%
             RAM:       54648 B       128 KB     41.69%
            ITCM:          0 GB       128 KB      0.00%
            DTCM:          0 GB       128 KB      0.00%
           OCRAM:      125764 B       768 KB     15.99%
        IDT_LIST:          0 GB         2 KB      0.00%
build finished with warning(s).

That was actually pretty painless, thanks @DerekSnell !
